editor/plugins/polygon_2d_editor_plugin.cpp

```diff
@@ -940,36 +946,86 @@ void Polygon2DEditor::_uv_input(const Ref<InputEvent> &p_input) {
 			uv_edit_draw->queue_redraw();
 		}
 	}
+}
 
-	Ref<InputEventMagnifyGesture> magnify_gesture = p_input;
-	if (magnify_gesture.is_valid()) {
-		uv_zoom->set_value(uv_zoom->get_value() * magnify_gesture->get_factor());
-	}
-
-	Ref<InputEventPanGesture> pan_gesture = p_input;
-	if (pan_gesture.is_valid()) {
-		uv_hscroll->set_value(uv_hscroll->get_value() + uv_hscroll->get_page() * pan_gesture->get_delta().x / 8);
-		uv_vscroll->set_value(uv_vscroll->get_value() + uv_vscroll->get_page() * pan_gesture->get_delta().y / 8);
+void Polygon2DEditor::_center_view() {
+	Size2 texture_size = Size2();
+	if (node->get_texture().is_valid()) {
+		texture_size = node->get_texture()->get_size();
+		Vector2 zoom_factor = (uv_edit_draw->get_size() - Vector2(1, 1) * 50 * EDSCALE) / texture_size;
+		zoom_widget->set_zoom(MIN(zoom_factor.x, zoom_factor.y));
+	} else {
+		zoom_widget->set_zoom(EDSCALE);
 	}
+	// Recalculate scroll limits.
+	_update_zoom_and_pan(false);
+
+	updating_zoom = true;
+	Size2 offset = (texture_size - uv_edit_draw->get_size() / uv_draw_zoom) / 2;
+	uv_hscroll->set_value(offset.x);
+	uv_vscroll->set_value(offset.y);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
 }
 
 void Polygon2DEditor::_uv_pan_callback(Vector2 p_scroll_vec, Ref<InputEvent> p_event) {
-	uv_hscroll->set_value(uv_hscroll->get_value() - p_scroll_vec.x);
-	uv_vscroll->set_value(uv_vscroll->get_value() - p_scroll_vec.y);
+	updating_zoom = true;
+	uv_hscroll->set_value(uv_hscroll->get_value() - p_scroll_vec.x / uv_draw_zoom);
+	uv_vscroll->set_value(uv_vscroll->get_value() - p_scroll_vec.y / uv_draw_zoom);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
 }
 
 void Polygon2DEditor::_uv_zoom_callback(float p_zoom_factor, Vector2 p_origin, Ref<InputEvent> p_event) {
-	uv_zoom->set_value(uv_zoom->get_value() * p_zoom_factor);
+	updating_zoom = true;
+	zoom_widget->set_zoom(uv_draw_zoom * p_zoom_factor);
+	uv_draw_ofs += p_origin / uv_draw_zoom - p_origin / zoom_widget->get_zoom();
+	uv_hscroll->set_value(uv_draw_ofs.x);
+	uv_vscroll->set_value(uv_draw_ofs.y);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
 }
 
-void Polygon2DEditor::_uv_scroll_changed(real_t) {
-	if (updating_uv_scroll) {
+void Polygon2DEditor::_update_zoom_and_pan(bool zoom_at_center) {
+	if (updating_zoom) {
 		return;
 	}
+	updating_zoom = true;
+	uv_draw_ofs = Vector2(uv_hscroll->get_value(), uv_vscroll->get_value());
+	real_t previous_zoom = uv_draw_zoom;
+	uv_draw_zoom = zoom_widget->get_zoom();
+	if (zoom_at_center) {
+		Vector2 center = uv_edit_draw->get_size() / 2;
+		uv_draw_ofs += center / previous_zoom - center / uv_draw_zoom;
+	}
 
-	uv_draw_ofs.x = uv_hscroll->get_value();
-	uv_draw_ofs.y = uv_vscroll->get_value();
-	uv_draw_zoom = uv_zoom->get_value();
+	Point2 min_corner = Point2();
```

Comment:
This will make the result never larger than `(0,0)`, is this intended?

Reply:
Yes, since the texture is being drawn in position (0,0), that should be always inside drawing area, plus it would be strange if you couldn't scroll towards origo.
This bit (L1005) might need to be revised if #84076 gets merged, since texture isn't always drawn at origo then.